### ISSUE TITLE
How to package multiple services into a single WAR file?

### ISSUE THREAD

**Issue by JohannBlake** [GoogleCloudPlatform/google-cloud-intellij#1912]
In Google App Engine (GAE) Standard, you can package multiple services into a single exploded EAR file but only if your app is running under Java 7. This does not work for Java 8:

> An App Engine application that uses services is organized as an
> unpacked Java Enterprise Archive (EAR) directory structure. The
> top-level EAR directory contains a single META-INF subdirectory, and a
> separate directory for each service in the app. Note that EAR is
> supported for the Java 7 runtime only, not for the Java 8 runtime.

I couldn't find anything in the GAE docs that indicates how you package multiple services into a single WAR file.

[https://cloud.google.com/appengine/docs/standard/java/configuration-files][1]

I realize that I can create a separate WAR file for each service but the local dev server supplied by Google for testing your app only accepts a single WAR file. I could run multiple instances of the dev server using different ports but that seems kind of messy.

  [1]: https://cloud.google.com/appengine/docs/standard/java/configuration-files

**Comment by patflynn** [comment]
@JohannBlake EAR support for App Engine Standard is on a path to deprecation and the new toolchain does not support it.  Most of our toolchain still supports running multiple services in the same devappserver run but sadly IntelliJ plugin does not (nor did it ever).

We also have instructions on setting this up for [Maven](https://cloud.google.com/appengine/docs/standard/java/tools/migrate-maven) and [Gradle](https://cloud.google.com/appengine/docs/standard/java/tools/migrate-gradle) users. Take a look at the 'Migrating EAR-based multi-service configurations' section of each guide.

@lesv multi-service configuration should probably be part of the main docset and not just the migration guide.

dupe of #1385

**Comment by JohannBlake** [comment]
@patflynn Thanks for the feedback. Unfortunately there is no documentation that indicates that your local dev server supports multiple services. If it does, can you please indicate how you are suppose to have it know where the app code is for each service? As the documentation currently stands, it only indicates that you are to provide the directory where the exploded WAR file is located. But an exploded WAR file cannot have multiple appengine-web.xml files. The documentation I am referring to is:

[https://cloud.google.com/appengine/docs/standard/java/tools/using-local-server](https://cloud.google.com/appengine/docs/standard/java/tools/using-local-server)

Even though IntelliJ doesn't support it, currently I run multiple instances of the local dev server, giving each a unique ip address. I do use IntelliJ and setup a remote connection for debugging and this does work. It does mean that I have to have each service opened up in a separate IntelliJ project window. Would be ideal if I could just open up the entire app with all services in a single IntelliJ project and only have a single instance of the local dev server running.

**Comment by patflynn** [comment]
Are you using Maven or Gradle?  If you take a look at the links I included
you'll see that the 'Migrating EAR-based multi-service configurations'
section of those docs describe how to do it with those build systems.

Here
<https://github.com/patflynn/appengine-potpourri/blob/cloud-sdk-refactor/endpoints-frontend/pom.xml>
is an example project that is configured to do so.

**Comment by JohannBlake** [comment]
I am using Gradle. When I clicked on the link you gave for Gradle, the title read:

**Migrating to the Cloud SDK-based Gradle plugin**

I skipped this because I was not migrating anything. I was building an app from scratch. Developers don't have time to read topics on migration if they are building new apps. I highly recommend that a link to the topic you gave be indicated on these two pages:

[https://cloud.google.com/appengine/docs/standard/java/tools/using-local-server](https://cloud.google.com/appengine/docs/standard/java/tools/using-local-server)

[https://cloud.google.com/appengine/docs/standard/java/configuration-files](https://cloud.google.com/appengine/docs/standard/java/configuration-files)

On the last link it states the following:

> An App Engine application that uses services is organized as an unpacked Java Enterprise Archive (EAR) directory structure. The top-level EAR directory contains a single META-INF subdirectory, and a separate directory for each service in the app. Note that EAR is supported for the Java 7 runtime only, not for the Java 8 runtime.

No mention is ever given on how to handle the configuration (or packaging) for a Java 8 app. I spent several hours googling to find the solution to Java 8 and without success.

I'll try out the instructions under migrating and if I have any problems, I'll get back to you. Thanks!

**Comment by JohannBlake** [comment]
I'm having a lot of difficulty getting gradle to run with the plugin for multiple modules. This really should not be surprising as there is not a single project anywhere, especially in github that actually shows how this is done. Google really needs to spend more time invested in providing skeleton projects that developers can use rather than us having to figure out how to make things work.

The biggest mistake Google constantly makes is assuming that every developer is a gradle expert. We are not. I have used gradle fairly extensively with Android but I am by no means an expert.

If you follow the instructions for the gradle plugin for GAE, it works fine if you are building a single project where the source code is directly off the root directory. For multiple modules, each module has its own build.gradle file and contains the settings as outlined in the docs for GAE.

So my project structure looks like this:

```
MyProject
  - Accounts // Folder for the accounts module (or accounts service)
  - Channels // Folder for the channels module (or channels service)
  build.gradle
  settings.gradle
```
  
  
The settings.gradle file contains:

> include 'Accounts', 'Channels'

The build.gradle file has the following:

```
buildscript {    // Configuration for building
    repositories {
        jcenter()    // Bintray's repository - a fast Maven Central mirror & more
        mavenCentral()
    }
    dependencies {
        classpath 'com.google.cloud.tools:appengine-gradle-plugin:+'    // latest App Engine Gradle tasks
    }
}

repositories {   // repositories for Jar's you access in your code
    maven {
        url 'https://oss.sonatype.org/content/repositories/snapshots' // SNAPSHOT repository (if needed)
    }
    mavenCentral()
    jcenter()
}

apply plugin: 'java'                              // standard Java tasks
apply plugin: 'com.google.cloud.tools.appengine'  // App Engine tasks

dependencies {
    compile 'com.google.appengine:appengine-api-1.0-sdk:+'  // Latest App Engine Api's
}

appengine {
    deploy {
        version = 'v1'
        project = 'my-project-id'
    }

    run {
        // configure the app to point to the right service directories
        services = [
                getExplodedAppDir(project),
                getExplodedAppDir(project(":Channels"))
        ]
    }
}

// helper method to obtain correct directory and set up dependencies
def getExplodedAppDir(Project p) {
    // if not 'this' module, then do some setup.
    if (p != project) {
        // make sure we evaluate other modules first so we get the right value
        evaluationDependsOn(p.path)
        // make sure we build "run" depends on the other modules exploding wars
        project.tasks.appengineRun.dependsOn p.tasks.explodeWar
    }
    return p.tasks.explodeWar.explodedAppDirectory
}

sourceCompatibility = 1.8     // App Engine Flexible uses Java 8
targetCompatibility = 1.8     // App Engine Flexible uses Java 8
```

But if I run just:

> gradle

it generates the error:

 

> Could not find method run() for arguments [build_6r74b64esmpnn8rfe3xbr081m$_run_closure3$_closure6@176878e2] on object of type com.google.cloud.tools.gradle.appengine.core.AppEngineExtension.

 
I don't have any issues when running gradle on any of the modules. This error only occurs when running gradle on the root project.

If you can see the issue, please indicate what is wrong. Or better yet, upload an actual skeleton project showing a working build.gradle file for multiple services.

**Comment by JohannBlake** [comment]
I was able to fix this problem by replacing:

`apply plugin: 'com.google.cloud.tools.appengine'`

with:

`apply plugin: "com.google.cloud.tools.appengine-standard"`

The build.gradle files for the modules actually use:

`apply plugin: 'com.google.cloud.tools.appengine'`

but the documentation at:

[https://github.com/GoogleCloudPlatform/app-gradle-plugin/blob/master/USER_GUIDE.md](https://github.com/GoogleCloudPlatform/app-gradle-plugin/blob/master/USER_GUIDE.md
)

states:

> You can also use the com.google.cloud.tools.appengine plugin that will automatically determine your environment based on the presence of an appengine-web.xml in src/main/webapp/WEB-INF/, Standard if present, Flexible otherwise.

But this does not seem to be the case. Using the com.google.cloud.tools.appengine plugin generates this bug.

Even after fixing this bug, I am now faced with a new problem:

> Task with path 'war' not found in root project 'GAE'.

Of course there is no path to WAR in the root directory. Each module generates its own output in its own build directory. So how am I suppose to fix this?

**Comment by patflynn** [comment]
@JohannBlake you should probably move this issue to the [app-gradle-plugin](https://github.com/GoogleCloudPlatform/app-gradle-plugin/issues) repo, although I suspect your problem is ultimately about inadequate documentation.

The [example app](https://github.com/patflynn/appengine-potpourri/) I shared with you appears to work fine on my machine:
`
git clone https://github.com/patflynn/appengine-potpourri.git
cd appengine-potpourri
./gradlew :endpoints-frontend:appengineRun
....
Please visit https://developers.google.com/appengine/downloads for the latest SDK.
********************************************************
Feb 15, 2018 1:28:23 PM com.google.apphosting.utils.jetty.JettyLogger info
INFO: Logging to JettyLogger(null) via com.google.apphosting.utils.jetty.JettyLogger
Feb 15, 2018 6:28:23 PM com.google.apphosting.utils.jetty.JettyLogger info
INFO: jetty-6.1.x
Feb 15, 2018 6:28:24 PM com.google.apphosting.utils.jetty.JettyLogger info
INFO: Started SelectChannelConnector@localhost:8080
Feb 15, 2018 6:28:24 PM com.google.appengine.tools.development.AbstractModule startup
INFO: Module instance default is running at http://localhost:8080/
Feb 15, 2018 6:28:24 PM com.google.appengine.tools.development.AbstractModule startup
INFO: The admin console is running at http://localhost:8080/_ah/admin
Feb 15, 2018 1:28:24 PM com.google.appengine.tools.development.DevAppServerImpl doStart
INFO: Dev App Server is now running
`

@loosebazooka FYI

**Comment by patflynn** [comment]
FWIW the instructions on the migration doc and the example I sent do not apply the appengine plugin to the project root, but instead to the primary/default module.  It's that primary module that should pull in the other services that you want to run.

**Comment by JohannBlake** [comment]
The link to the sample that you provided:

[https://github.com/patflynn/appengine-potpourri/](https://github.com/patflynn/appengine-potpourri/)

does not have any build.gradle files that use the setup for the gradle plugin as described at:

[https://cloud.google.com/appengine/docs/standard/java/tools/migrate-gradle](https://cloud.google.com/appengine/docs/standard/java/tools/migrate-gradle)

or at:

[https://github.com/GoogleCloudPlatform/app-gradle-plugin/blob/master/USER_GUIDE.md](https://github.com/GoogleCloudPlatform/app-gradle-plugin/blob/master/USER_GUIDE.md)

More specifically, it doesn't use the following code:

```
appengine {
    deploy {
        version = '1'
        project = "positive-karma-844"
    }

    run {
        services = [
                getExplodedAppDir(project),
                getExplodedAppDir(project(":{your-secondary-service}"))
        ]
    }
}

// helper method to obtain correct directory and set up dependencies
def getExplodedAppDir(Project serverProject) {
    // if not 'this' module, then do some setup.
    if (serverProject != project) {
        // make sure we evaluate other modules first so we get the right value
        evaluationDependsOn(serverProject.path)
        // make sure we build "run" depends on the other modules exploding wars
        project.tasks.appengineRun.dependsOn serverProject.tasks.explodeWar
    }
    return serverProject.tasks.explodeWar.explodedAppDirectory
}
```

I hope that you can understand the frustration I've gone through.

**Comment by patflynn** [comment]
Ah I missed specifying the branch the second time I sent you the link. Make
sure you're looking at the 'cloud-sdk-refactor' branch:
https://github.com/patflynn/appengine-potpourri/blob/cloud-sdk-refactor/endpoints-frontend/build.gradle

**Comment by JohannBlake** [comment]
Thanks. That did it. I also ran the sample and it runs as expected. Now I have something to compare my app to in order to find out why mine doesn't work. It would be helpful to other developers if that sample project was more obviously available on the GCP doc site. Being buried in some branch on github isn't of much help. All the same, thanks for helping out!

**Comment by JohannBlake** [comment]
A minor issue I came across testing the sample app you gave me the link to. When I open up the browser and navigate to the path:

`http://localhost:49829/_ah/admin`

and then click on the "modules" link, it generates an error:

`java.lang.NullPointerException: null value in entry: version=null`

Does this happen on your machine? Am I not suppose to be expecting the two services to be listed?

Thanks.

**Comment by patflynn** [comment]
I think that may be fixed if you have the latest cloud SDK version.  If not
specifying <version/> in the appengine-web.xml resolves it (I think).

**Comment by JohannBlake** [comment]
I have the latest version of the cloud sdk but that didn't fix it. Putting the version number in the appengine-web.xml did.

I have a related question to all of this. When I view the services with http://localhost:49829/_ah/admin

it is clear that each is reachable under a different port. That was clear from the docs and is expected. What isn't clear is how devs are suppose to handle this in code when debugging locally. For example, if I write code in one service to call a different service, the http url would have to be:

```
http://localhost:<port-service-a>/...
http://localhost:<port-service-b>/...
http://localhost:<port-service-c>/...
```

Each service has its own port and these get randomly generated each time you run:

`gradle appengineRun`

Obviously it makes no sense to hardcode ports in the url and have to change it each time you run the app. There is a port number that can be specified in the gradle plugin but that is only a single port number and presumably applies to the default service. So how do you handle ports that change each time the app is run and you need to communicate from one service to the other?

**Comment by patflynn** [comment]
@JohannBlake I don't know the best practices for this but I'm pretty sure you're not supposed to try and address services directly using the hostname port.  [Dispatch.xml ](https://cloud.google.com/appengine/docs/standard/java/config/dispatchref) or taskqueues could be a way of solving this but I'm not sure.

@lesv FYI

Here's a similar question on [Stackoverflow](https://stackoverflow.com/questions/42097508/communication-methods-between-two-gae-services-in-the-same-project).

**Comment by JohannBlake** [comment]
Unfortunately the "best practice" was never addressed in the docs located at:

[https://cloud.google.com/appengine/docs/standard/java/how-requests-are-routed](https://cloud.google.com/appengine/docs/standard/java/how-requests-are-routed)

**Comment by lesv** [comment]
@patflynn Yep - we should incorporate this whole thread.

**Comment by briandealwis** [comment]
@JohannBlake 

One option is to use the App Engine API's [_ModulesService_](https://cloud.google.com/appengine/docs/standard/java/javadoc/com/google/appengine/api/modules/ModulesService) to get local _host:port_ connection points for your services (aka modules).  The full _ModulesService_ API isn't supported, but you can use:

  - `getCurrentModule()`
  - `getCurrentVersion()`: always returns `no_version`
  - `getModules()`: returns the service IDs
  - `getVersionHostname(serviceId, "no_version")`: returns `localhost:XXXX` on the local devappserver, and returns the dot-form hostname to reach the service on App Engine (e.g., _version.serviceId.projectId.appspot.com_)

If you deploy your services with same versions (e.g., `blue` or `green` if using a blue-green deployment strategy), using the _ModulesService_ would be really easy since you'd always be accessing service versions that corresponds to the current service's version.

If you deploy using different versions, you must have some kind of service-version table, in which case you could special case the [local-run](https://cloud.google.com/appengine/docs/standard/java/tools/using-local-server#detecting_the_application_runtime_environment) situation to always use `no_version`.

**Comment by JohannBlake** [comment]
I created a github project that goes into details on how to setup and debug microservices for GAE in a Java project using IntelliJ. The solution I have provided avoids the need to use localhost:port and lets you debug your app with almost the same domain name that is used in GAE with the only difference being that on GAE you would normally use https and no port is specified, whereas locally you would use http and specify a port. Other than these minor differences, the url's are identical.

There is no need to use the gradle plugin for running locally although it is still used for deploying to GAE. You just don't need it to run the app, meaning no need for:

> gradle appengineRun

Using the same endpoints locally as they would appear in GAE also allows you to use the same code for inspecting the subdomain and path, should that information be required by the servlet.

The github project can be found at:

[https://github.com/JohannBlake/gae-microservices](https://github.com/JohannBlake/gae-microservices)
